### PR TITLE
Add typing to all calls to ``self.stats``

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

All checkers inherit from a baseclass which has a ``stats`` attribute.
This attribute has a fairly unmanageable type, but the current typing includes all variations of the attribute.
Other changes not directly related to ``self.stats`` are due to ``mypy`` warnings.
This incorporate the feedback received in #4954

### CODE REVIEW

File: pylint/reporters/multi_reporter.py
Comment (DanielNoord) [inline pylint/reporters/multi_reporter.py:98]: I changed this to `Dict` instead of `Mapping` as its only ever called with `self.stats` and using `Mapping` gave a `Mypy` warning because of an incorrect argument type for one of the calls to `on_close`.

**Review by Pierre-Sassoulas — COMMENTED** [review]
I was thinking about creating a class so the typing is simpler and we can unit test it easily.

File: pylint/checkers/__init__.py
Comment (DanielNoord) [inline pylint/checkers/__init__.py:59]: The `List` included in the type of all `stats` comes from https://github.com/PyCQA/pylint/blob/40cc2ffd7887959157aaf469e09585ec2be7f528/pylint/checkers/imports.py#L469
This is then used in https://github.com/PyCQA/pylint/blob/40cc2ffd7887959157aaf469e09585ec2be7f528/pylint/graph.py#L170-L182

File: pylint/checkers/__init__.py
Comment (cdce8p) [inline pylint/checkers/__init__.py:49]: `typing.Counter` was added in `3.6.1`. Thus to use it, we need to add a guard unfortunately.
https://docs.python.org/3/library/typing.html#typing.Counter

File: pylint/checkers/__init__.py
Comment (cdce8p) [inline pylint/checkers/__init__.py:63]: I'm a bit surprised that the code works with so many different types.
@Pierre-Sassoulas mentioned that we should probably create a `Stats` class or something like it, I agree. However as it would be a larger refactoring for another time, maybe a TypeAlias will work for now.

I would suggest to create `pylint/typing.py` as a helper file to store TypeAliases, Enums, NamedTuple, TypedDicts, and so on. (Only if it doesn't make sense somewhere else of course.)

**Review by cdce8p — CHANGES_REQUESTED** [review]
Not a full review just yet

File: pylint/checkers/__init__.py
Comment (DanielNoord) [reply pylint/checkers/__init__.py]: I borrowed the solution from:
https://github.com/PyCQA/pylint/blob/40cc2ffd7887959157aaf469e09585ec2be7f528/pylint/checkers/strings.py#L51-L52
Let me know if this is correct!

File: tests/pyreverse/test_writer.py
Comment (DanielNoord) [inline tests/pyreverse/test_writer.py:31]: For some reason the import of `conftest` introduced in https://github.com/PyCQA/pylint/commit/25000863761220c1aa5db129e3cd41f5fe51b167 failed the `pre-commit`.

File: pylint/checkers/__init__.py
Comment (cdce8p) [inline pylint/checkers/__init__.py:77]: This should probably be `lines.extend( [ ... ] )`, not `lines +=`

File: pylint/checkers/base.py
Comment (cdce8p) [inline pylint/checkers/base.py:422]: This should be typed implicitly.

File: pylint/checkers/raw_metrics.py
Comment (cdce8p) [inline pylint/checkers/raw_metrics.py:38]: Again, probably not necessary.

File: pylint/lint/parallel.py
Comment (cdce8p) [inline pylint/lint/parallel.py:6]: Unguarded `Counter` import.

File: pylint/lint/parallel.py
Comment (cdce8p) [inline pylint/lint/parallel.py:37]:
```suggestion
    by_msg: "Counter[str]" = collections.Counter()
```

File: pylint/lint/parallel.py
Comment (cdce8p) [inline pylint/lint/parallel.py:39]:
```suggestion
        message_stats: Union["Counter", Dict] = stat.pop("by_msg", {})  # type: ignore
```

File: pylint/lint/report_functions.py
Comment (cdce8p) [inline pylint/lint/report_functions.py:41]: Again, implicit.

File: pylint/typing.py
Comment (cdce8p) [inline pylint/typing.py:14]: I'm wondering: Should this be `CheckerStats` instead? It's almost always `self.stats`, so it would make sense to use it here as well.

File: tests/pyreverse/test_writer.py
Comment (cdce8p) [reply tests/pyreverse/test_writer.py:31]: Shouldn't this be `from pylint.pyrevers.conftest import PyreverseConfig`?

**Review by cdce8p — COMMENTED** [review]
I still feel like I need to think about this a bit longer. Left some comments nevertheless.

File: tests/pyreverse/test_writer.py
Comment (DanielNoord) [reply tests/pyreverse/test_writer.py:31]: See discussion here: https://github.com/PyCQA/pylint/pull/4950#discussion_r702048576
Seems like we can't fix this without a (big) refactor

File: pylint/checkers/base.py
Comment (DanielNoord) [reply pylint/checkers/base.py:422]: The problem is on line 437. Because the length of the tuple changes there. `mypy` doesn't like that and complains about a difference in typing. I could add an ignore on that line (and on the other cases you mentioned), but `Tuple[str, ...]` might be better in this case.

File: pylint/checkers/raw_metrics.py
Comment (DanielNoord) [reply pylint/checkers/raw_metrics.py:38]: See above

File: pylint/lint/report_functions.py
Comment (DanielNoord) [reply pylint/lint/report_functions.py:41]: Similar problem, this time on line 43.
https://github.com/PyCQA/pylint/blob/04706af5d411ef5359fa195565f21d3fef7aed2b/pylint/lint/report_functions.py#L41-L43

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas You might want to change the milestone on this (or other PRs). Same goes for #4980 in order for the current typing efforts of you and me not to become double-work.

Just to let you know, I didn't plan on opening any new typing PRs before the current ones are all merged. Probably best to work on getting the current ones merged first. Perhaps you could help me with a review of #4980 as it requires quite extensive knowledge of `pylint` 😄

File: pylint/checkers/base.py
Comment (cdce8p) [reply pylint/checkers/base.py:422]: If additional items are added later, wouldn't it be better if lines where a `List[str]` instead?

File: pylint/lint/parallel.py
Comment (cdce8p) [inline pylint/lint/parallel.py:14]: This could also be a `sys.version_info` guard with import from `typing_extensions`. In that case the quotes around `Counter` wouldn't be necessary. Leave that up to you, if you what to change it. Support for `3.6` will end soon anyway, so might not be that important.

File: pylint/reporters/reports_handler_mix_in.py
Comment (cdce8p) [inline pylint/reporters/reports_handler_mix_in.py:78]: Can `**kwargs` be typed? I.e. do you know the type for the dict values (in `kwargs`)?

File: pylint/reporters/ureports/nodes.py
Comment (cdce8p) [inline pylint/reporters/ureports/nodes.py:19]:
```suggestion
from typing import Optional
```
The empty lines are not really needed.

File: pylint/typing.py
Comment (cdce8p) [inline pylint/typing.py:5]:
```suggestion
"""A collection of typing utilities."""
```

File: pylint/typing.py
Comment (cdce8p) [inline pylint/typing.py:7]:
```suggestion
from typing import TYPE_CHECKING, Dict, List, Union
```
Empty line

File: pylint/typing.py
Comment (cdce8p) [inline pylint/typing.py:10]: If you decide to change it, this should be changed as well.

File: pylint/checkers/base.py
Comment (DanielNoord) [reply pylint/checkers/base.py:422]: Yes! I can make all occurrences of `lines` a list and type them as `List[str]`. Want me to do that?
Originally I didn't want to change any code if not necessary and only add typing.

File: pylint/lint/parallel.py
Comment (DanielNoord) [reply pylint/lint/parallel.py:14]: Do you know if `typing` imports get skipped during runtime? If not it might be better to keep it under the `if TYPE_CHECKING`?

File: pylint/reporters/reports_handler_mix_in.py
Comment (DanielNoord) [reply pylint/reporters/reports_handler_mix_in.py:78]: Seems to be a mix of `int`, `list`, `dict` and `None`... I guess not? Or I could add a `Union`. What do you prefer?

File: pylint/checkers/base.py
Comment (cdce8p) [reply pylint/checkers/base.py:422]: I think that would make sense. Could you do that in a separate PR? Just the change to `list[str]`. That should be easy to review and fast to merge.

Comment (DanielNoord) [reply pylint/checkers/base.py:422]: Yes, give me a couple of minutes!

Comment (cdce8p) [reply pylint/checkers/base.py:422]: Tag me, and I'll take a look.

File: pylint/lint/parallel.py
Comment (cdce8p) [reply pylint/lint/parallel.py:14]: No, they are not skipped. Basically they are normal imports. However, since `typing` as well as `typing_extensions` are already imported elsewhere, it shouldn't have much of a performance impact. If any at all.

File: pylint/reporters/reports_handler_mix_in.py
Comment (cdce8p) [reply pylint/reporters/reports_handler_mix_in.py:78]: With that mix it might be a case for `Any` 🤔

File: pylint/checkers/base.py
Comment (DanielNoord) [reply pylint/checkers/base.py:422]: See https://github.com/PyCQA/pylint/pull/5009

File: pylint/lint/parallel.py
Comment (DanielNoord) [reply pylint/lint/parallel.py:14]: Hmm okay. If you don't have a preference for either I think I will leave it at this.
I actually was thinking about making a checker for imports only being used in typing and a `if TYPE_CHECKING` statement. So this fits nicely!

File: pylint/reporters/reports_handler_mix_in.py
Comment (DanielNoord) [reply pylint/reporters/reports_handler_mix_in.py:78]: Although `mypy` will probably complain about that in the future... I will add the Union 😄

File: pylint/lint/parallel.py
Comment (cdce8p) [reply pylint/lint/parallel.py:14]: > I actually was thinking about making a checker for imports only being used in typing and a `if TYPE_CHECKING` statement. So this fits nicely!

Just a gut feeling: Don't know how useful such a checker would really be.

File: pylint/reporters/reports_handler_mix_in.py
Comment (Pierre-Sassoulas) [reply pylint/reporters/reports_handler_mix_in.py:78]: I kinda agree with Marc here, this much type in a Union scream "whatever" or as mypy call it ``Any`` 😄 We'd need to refactor to make this code make sense and we'll probably throw some code if we create a proper class for stats later so we might as well not bother with too much precision in unsalvageable legacy code.

Comment (cdce8p) [reply pylint/reporters/reports_handler_mix_in.py:78]: > We'd need to refactor to make this code make sense and we'll probably throw some code if we create a proper class for stats later so we might as well not bother with too much precision in unsalvageable legacy code.

Good point! Although imprecise, sometimes `Any` is the better choice after all. Even if only temporarily.

File: pylint/lint/parallel.py
Comment (DanielNoord) [reply pylint/lint/parallel.py:14]: I thought it might be useful for Python 3.9, where people are more likely to not import from `typing`. For example, in some of my own projects I tend to quickly import from a dependency when I am working in a file that only has a helper function. Although performance effects might be negligible, (I think) it fits with the idea of `pylint` suggesting best practice.
But, this will probably be a discussion for when I make the first PR for this checker 😄

File: pylint/reporters/reports_handler_mix_in.py
Comment (DanielNoord) [reply pylint/reporters/reports_handler_mix_in.py]: Good points! Add `Any` in latest commit!

File: pylint/checkers/__init__.py
Comment (cdce8p) [inline pylint/checkers/__init__.py:61]: This could also be `Iterable[str]`. Just a suggestion, not strictly necessary.

**Review by cdce8p — COMMENTED** [review]
@Pierre-Sassoulas Would you mind taking a look at this? Overall I think it's good. Haven't checked each type tough.

**Review by Pierre-Sassoulas — APPROVED** [review]
I like the new class name, and this cleans up a lot of potential problem too 👍

**Comment by Pierre-Sassoulas** [comment]
> Shouldn't this be from pylint.pyrevers.conftest import PyreverseConfig?

Yes, I opened #5010

**Pull request merged**
